tests/test_update_picking.py

```diff
@@ -0,0 +1,503 @@
+                            },
+                        ]
+        picking.update_picking(products_info=products_info)
+        cherry_move_lines = picking.move_line_ids.filtered(lambda x: x.product_id == cherry)
+        apple_move_lines = picking.move_line_ids.filtered(lambda x: x.product_id == self.apple)
+        self.assertEqual(sum(cherry_move_lines.mapped('qty_done')), 2)
+        self.assertEqual(apple_move_lines.qty_done, 2)
+        picking.update_picking(validate=True)
+        self.assertEqual(picking.state, 'done', 'Process doesn\'t complete')
+
+    def test16_update_picking_set_result_package_id(self):
+        """ checks the correct result_package_id is set when
+            result_package_name is used in update_picking
+        """
+        Package = self.env['stock.quant.package']
+        package = Package.get_package('test_package', create=True)
+
+        creation_info = [{'product': self.apple, 'qty': 2}]
+        picking = self.create_picking(self.picking_type_in,
+                                     origin='test_picking_origin',
+                                     products_info=creation_info,
+                                     confirm=True)
+
+        products_info = [{'product_barcode': self.apple.barcode, 'qty':2}]
+        picking.update_picking(products_info=products_info, result_package_name=package.name)
+        self.assertEqual(picking.move_line_ids.result_package_id, package)
+
+        picking.update_picking(validate=True)
+        self.assertEqual(picking.state, 'done', 'Process doesn\'t complete')
+
+    def test17_update_picking_set_result_package_id_mixed_package(self):
+        """ checks the correct result_package_id is set when
+            result_package_name is used in update_picking when the
+            there is two products
+        """
+        Package = self.env['stock.quant.package']
+        package = Package.get_package('test_package', create=True)
+
+        cherry = self.create_product('Cherry')
+        creation_info = [
+                            {'product': self.apple, 'qty': 2},
+                            {'product': cherry, 'qty': 2},
+                        ]
+        picking = self.create_picking(self.picking_type_in,
+                                     origin='test_picking_origin',
+                                     products_info=creation_info,
+                                     confirm=True)
+        products_info = [
+                            {
+                                'product_barcode': self.apple.barcode,
+                                'qty':2
+                            },
+                            {
+                                'product_barcode': cherry.barcode,
+                                'qty': 2,
+                            },
+                        ]
+        picking.update_picking(products_info=products_info, result_package_name=package.name)
+        self.assertEqual(picking.mapped('move_line_ids.result_package_id'), package)
```

Comment:
If only one is set, I think it will pass the test. What about changing it to something like:
`self.assertEqual(len(picking.mapped('move_line_ids').filtered(lambda ml: ml.result_package_id == package)), 2)`